controllers/services/auth/auth_controller_actions.go

```diff
@@ -0,0 +1,231 @@
+		}
+	}
+
+	return nil
+}
+
+func bindRole(ctx context.Context, rr *odhtypes.ReconciliationRequest, groups []string, roleBindingName string, roleName string) error {
+	var groupsToBind []rbacv1.Subject
+	for _, e := range groups {
+		rs := rbacv1.Subject{
+			Kind:     "Group",
+			APIGroup: "rbac.authorization.k8s.io",
+			Name:     e,
+		}
+		groupsToBind = append(groupsToBind, rs)
+	}
+
+	rb := &rbacv1.RoleBinding{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      roleBindingName,
+			Namespace: rr.DSCI.Spec.ApplicationsNamespace,
+		},
+		Subjects: groupsToBind,
+		RoleRef: rbacv1.RoleRef{
+			APIGroup: "rbac.authorization.k8s.io",
+			Kind:     "Role",
+			Name:     roleName,
+		},
+	}
+	err := rr.Client.Update(ctx, rb)
+	if err != nil {
+		return errors.New("error creating RoleBinding for group")
+	}
+
+	return nil
+}
+
+func removeOldPermissions(ctx context.Context, rr *odhtypes.ReconciliationRequest, groups []string, roleBindingName string, roleName string) error {
+	roleBinding := &rbacv1.RoleBinding{}
+	if err := rr.Client.Get(ctx, client.ObjectKey{
+		Name:      roleBindingName,
+		Namespace: rr.DSCI.Spec.ApplicationsNamespace,
+	}, roleBinding); err != nil {
+		return client.IgnoreNotFound(err)
+	}
+
+	var groupsToRemove []rbacv1.Subject
+	for _, s := range roleBinding.Subjects {
+		var found = false
+		for _, g := range groups {
+			if s.Name == g {
+				found = true
+			}
+		}
+		if !found {
+			groupsToRemove = append(groupsToRemove, s)
+		}
+	}
+
+	if len(groupsToRemove) > 0 {
+		rb := &rbacv1.RoleBinding{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      roleBindingName,
+				Namespace: rr.DSCI.Spec.ApplicationsNamespace,
+			},
+			Subjects: groupsToRemove,
+			RoleRef: rbacv1.RoleRef{
+				APIGroup: "rbac.authorization.k8s.io",
+				Kind:     "Role",
+				Name:     roleName,
+			},
+		}
+
+		err := rr.Client.Delete(ctx, rb)
+		if err != nil {
+			return errors.New("error deleting RoleBinding for group")
+		}	
+	}
+
+	return nil
+}
+
+func managePermissions(ctx context.Context, rr *odhtypes.ReconciliationRequest) error {
+	ai, ok := rr.Instance.(*serviceApi.Auth)
+	if !ok {
+		return errors.New("instance is not of type *services.Auth")
+	}
+
+	for _, role := range rr.Resources {
+		err := rr.Client.Update(ctx, &role)
+		if err != nil {
+			return err
+		}
+	}
+
+	bindRole(ctx, rr, ai.Spec.AdminGroups, "admingroup-rolebinding", "admingroup-role")
+	bindRole(ctx, rr, ai.Spec.AllowedGroups, "allowedgroup-rolebinding", "allowedgroup-role")
+
+	removeOldPermissions(ctx, rr, ai.Spec.AdminGroups, "admingroup-rolebinding", "admingroup-role")
+	removeOldPermissions(ctx, rr, ai.Spec.AllowedGroups, "allowedgroup-rolebinding", "allowedgroup-role")
+
+	return nil
+}
+
+func setStatus(ctx context.Context, rr *odhtypes.ReconciliationRequest) error {
+	ai, ok := rr.Instance.(*serviceApi.Auth)
+	if !ok {
+		return errors.New("instance is not of type *services.Auth")
+	}
+
+	ai.Status.ObservedGeneration = ai.Status.ObservedGeneration+1
```

Comment:
usually `ObservedGeneration`  should mirror the `spec.Generation`, it is meant to represent the spec generation the controller is acting upon.

Reply:
Now I think on it, i'd be more inclined to remove the observedGeneration. The CR content is straightforward and not expected to change very often right now. Maybe computing an observedGeneration can wait for the complexity to requre it.

Reply:
I'd keep it, just copy the `spec.Generation`, it helps detecting if your controller is stuck somewhere when you are making changes to the spec

Reply:
I don't see any spec generation, do you mean metadata generation?

Reply:
yep, that one (sorry for the confusion)

Reply:
:+1:  I've made that change now.